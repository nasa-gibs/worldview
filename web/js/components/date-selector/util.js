import util from '../../util/util';

export const monthStringArray = [
  'JAN',
  'FEB',
  'MAR',
  'APR',
  'MAY',
  'JUN',
  'JUL',
  'AUG',
  'SEP',
  'OCT',
  'NOV',
  'DEC',
];

/**
 *
 * @param {*} value
 * @param {*} date
 * @returns
 */
export const yearValidation = (value, dateParam, validate) => {
  const date = new Date(dateParam);
  if (value > 1000 && value < 9999) {
    const newDate = new Date(date.setUTCFullYear(value));
    return validate(newDate);
  }
  return null;
};

/**
 *
 * @param {*} value
 * @param {*} date
 * @returns
 */
export const monthValidation = (value, dateParam, validate) => {
  const date = new Date(dateParam);
  let newDate;
  // eslint-disable-next-line no-restricted-globals
  if (!isNaN(value) && value < 13 && value > 0) {
    newDate = new Date(date.setUTCMonth(value - 1));
    if (newDate) {
      return validate(newDate);
    }
    return null;
  }
  const realMonth = util.stringInArray(monthStringArray, value);
  if (realMonth !== false) {
    const day = date.getUTCDate();
    const zeroDay = new Date(date.setUTCDate(1));

    const zeroAddMonth = new Date(zeroDay.setUTCMonth(realMonth));
    const zeroAddedMonthNumber = zeroAddMonth.getUTCMonth();

    const addDay = new Date(zeroAddMonth.setUTCDate(day));
    const addedDayMonthNumber = addDay.getUTCMonth();

    if (addedDayMonthNumber !== zeroAddedMonthNumber) {
      return false;
    }
    return validate(addDay);
  }
  return null;
};

/**
 *
 * @param {*} value
 * @param {*} date
 * @returns
 */
export const dayValidation = (value, dateParam, validate) => {
  const date = new Date(dateParam);
  const standardMaxDateForMonth = 31;

  if (value > 0 && value <= standardMaxDateForMonth) {
    const actualMaxDateForMonth = new Date(
      date.getYear(),
      date.getMonth() + 1,
      0,
    ).getDate();

    if (value > actualMaxDateForMonth) {
      return false;
    }
    const newDate = new Date(date.setUTCDate(value));
    return validate(newDate);
  }
  return null;
};

/**
 *
 * @param {*} value
 * @param {*} date
 * @returns
 */
export const hourValidation = (value, dateParam, validate) => {
  const date = new Date(dateParam);
  if (value >= 0 && value <= 23) {
    const newDate = new Date(date.setUTCHours(value));
    return validate(newDate);
  }
  return null;
};

/**
 *
 * @param {*} value
 * @param {*} date
 * @returns
 */
export const minuteValidation = (value, dateParam, validate) => {
  const date = new Date(dateParam);
  if (value >= 0 && value <= 59) {
    const newDate = new Date(date.setUTCMinutes(value));
    return validate(newDate);
  }
  return null;
};
