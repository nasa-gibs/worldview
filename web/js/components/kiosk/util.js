// formats date for kiosk mode and updates to EST
// eslint-disable-next-line import/prefer-default-export
export function formatKioskDate(date, subdaily) {
  const options = {
    year: 'numeric',
    month: 'long',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    hour12: false,
    timeZone: 'America/New_York',
  };

  const formatter = new Intl.DateTimeFormat('en-US', options);
  const dateParts = formatter.formatToParts(date);

  const year = dateParts.find((part) => part.type === 'year').value;
  const month = dateParts.find((part) => part.type === 'month').value.slice(0, 3);
  const day = dateParts.find((part) => part.type === 'day').value;
  const hours = dateParts.find((part) => part.type === 'hour').value;
  const minutes = dateParts.find((part) => part.type === 'minute').value;

  const formattedDate = subdaily ? `${day} ${month} ${year} ${hours}:${minutes}:00 EDT` : `${day} ${month} ${year}`;

  return formattedDate;
}


