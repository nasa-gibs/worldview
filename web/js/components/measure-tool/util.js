import {
  MultiLineString as OlGeomMultiLineString,
  Polygon as OlGeomPolygon,
} from 'ol/geom';
import geographiclib from 'geographiclib';

const geod = geographiclib.Geodesic.WGS84;
const geographicProj = 'EPSG:4326';

/**
 * Shift x value of every coord except the last,
 * which is the active drawing point
 * @param {*} coords
 * @param {*} shiftValue
 */
function shiftXCoords (coords, shiftValue) {
  const newCoordsArray = [];
  const len = coords.length;
  for (let i = 0; i < len - 1; i += 1) {
    const [x, y] = coords[i];
    newCoordsArray.push([x + shiftValue, y]);
  }
  newCoordsArray.push(coords[len - 1]);
  return newCoordsArray;
}

/**
 * If the last two coordinates in a measurement are more than 180
 * degrees apart, flip the whole drawing forward or backward 360 degrees
 * @param {*} geom
 */
function checkForXFlip(geom, projection) {
  const coords = geom.getCoordinates();
  const [x1] = coords[coords.length - 2];
  const [x2] = coords[coords.length - 1];
  if (Math.abs(x1 - x2) > 180 && projection === geographicProj) {
    if (x1 < x2) {
      geom.setCoordinates(shiftXCoords(coords, 360));
    } else if (x1 > x2) {
      geom.setCoordinates(shiftXCoords(coords, -360));
    }
  }
  return geom;
}

/**
 * Transforms a LineString of two points to a MultiLineString of multiple points
 * applying a great circle arc transformation
 * @param {*} geom - the geometry object to apply great circle arc transformation to
 */
export function transformLineStringArc(geom, projection) {
  const coords = [];
  const distance = 10000; // meters between segments
  const transformedGeom = checkForXFlip(geom, projection).clone().transform(projection, geographicProj);
  transformedGeom.forEachSegment((segStart, segEnd) => {
    const line = geod.InverseLine(segStart[1], segStart[0], segEnd[1], segEnd[0]);
    const n = Math.ceil(line.s13 / distance);
    for (let i = 0; i <= n; i += 1) {
      const s = Math.min(distance * i, line.s13);
      const r = line.Position(s, geographiclib.Geodesic.LONG_UNROLL);
      coords.push([r.lon2, r.lat2]);
    }
  });
  return new OlGeomMultiLineString([coords]).transform(geographicProj, projection);
}

/**
 * Transforms a Polygon into one with addiitonal points on each edge to account for
 * great circle arc
 * @param {*} geom - the geometry object to apply great circle arc transformation to
 */
export function transformPolygonArc(geom, projection) {
  const coords = [];
  const transformedGeom = geom.clone().transform(projection, geographicProj);
  const distance = 10000; // meters between segments
  const polyCoords = transformedGeom.getCoordinates()[0];
  for (let i = 0; i < polyCoords.length - 1; i += 1) {
    const line = geod.InverseLine(
      polyCoords[i][1],
      polyCoords[i][0],
      polyCoords[i + 1][1],
      polyCoords[i + 1][0],
    );
    const n = Math.ceil(line.s13 / distance);
    for (let j = 0; j <= n; j += 1) {
      const s = Math.min(distance * j, line.s13);
      const r = line.Position(s, geographiclib.Geodesic.LONG_UNROLL);
      coords.push([r.lon2, r.lat2]);
    }
  }
  return new OlGeomPolygon([coords]).transform(geographicProj, projection);
}

/**
 * Calculate area of a polygon with GeographicLib library
 * @param {*} polygon
 * @returns {Number} - area in square meters
 */
export function getGeographicLibArea(polygon) {
  const coordinates = polygon.getCoordinates()[0];
  if (coordinates.length < 3) return 0;
  const geoPoly = geod.Polygon(false);
  coordinates.forEach((coord) => {
    // flip lat/lon position
    geoPoly.AddPoint(coord[1], coord[0]);
  });
  const { area } = geoPoly.Compute(false, true);
  return Math.abs(area);
}

/**
 * Calculate distance of a line with GeographicLib library
 * @param {*} polygon
 * @returns {Number} - distance in meters
 */
export function getGeographicLibDistance(line) {
  let totalDistance = 0;
  line.forEachSegment((segStart, segEnd) => {
    const r = geod.Inverse(segStart[1], segStart[0], segEnd[1], segEnd[0]);
    totalDistance += r.s12;
  });
  return totalDistance;
}
