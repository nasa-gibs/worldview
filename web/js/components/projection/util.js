export function parse(state, errors, config) {
  // Permalink version 1.0 - 1.1
  if (state.switch) {
    state.p = state.switch;
    delete state.switch;
  }
  var projId = state.p;
  if (projId) {
    if (!config.projections[projId]) {
      delete state.p;
      errors.push({
        message: 'Unsupported projection: ' + projId
      });
    }
  }
}
