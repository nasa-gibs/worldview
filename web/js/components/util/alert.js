import React from 'react';
import PropTypes from 'prop-types';
import { Alert } from 'reactstrap';
import { Portal } from 'react-portal';
/*
 * A react component, Builds a rather specific
 * interactive widget
 *
 * @class AlertComponent
 * @extends React.Component
 */
export default class AlertComponent extends React.Component {
  constructor(props) {
    super(props);
    if (props.timeout && props.onDismiss) {
      this.timeout = setTimeout(() => {
        props.onDismiss();
      }, props.timeout);
    }
  }

  componentDidMount() {
    this.mounted = true;
  }

  componentWillUnmount() {
    if (this.timeout) {
      clearTimeout(this.timeout);
    }
  }

  render() {
    return (
      <Portal node={document && document.getElementById('wv-content')}>
        <Alert className="wv-alert" isOpen={this.props.isOpen}>
          <div
            className="alert-content"
            title={this.props.title}
            onClick={this.props.onClick}
          >
            <span>
              {this.props.iconClassName ? (
                <i className={this.props.iconClassName} />
              ) : (
                ''
              )}
              {this.props.message}
              {this.props.onClick ? (
                <span>
                  &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Read More...
                </span>
              ) : null}
            </span>
          </div>
          <div className="close-alert" onClick={this.props.onDismiss}>
            <i className="fa fa-times exit fa-1x" />
          </div>
        </Alert>
      </Portal>
    );
  }
}
AlertComponent.defaultProps = {
  iconClassName: 'fas fa-exclamation-triangle fa-1x',
  title: ''
};
AlertComponent.propTypes = {
  iconClassName: PropTypes.string,
  isOpen: PropTypes.bool,
  message: PropTypes.string,
  onClick: PropTypes.func,
  onDismiss: PropTypes.func,
  timeout: PropTypes.number,
  title: PropTypes.string
};
