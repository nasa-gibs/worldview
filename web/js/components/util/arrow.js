import React from 'react';
import PropTypes from 'prop-types';

/*
 * @function Arrow Up/Down
 */
function Arrow({
  onClick, type, direction, isKioskModeActive,
}) {
  const containerClassName = `date-arrows date-arrow-${direction}`;
  const arrowClassName = `${direction}arrow`;

  return (
    <div
      onClick={onClick}
      className={isKioskModeActive ? 'd-none' : containerClassName}
      data-interval={type}
    >
      <svg width="25" height="8">
        <path d="M 12.5,0 25,8 0,8 z" className={arrowClassName} />
      </svg>
    </div>
  );
}

Arrow.propTypes = {
  direction: PropTypes.string,
  onClick: PropTypes.func,
  type: PropTypes.string,
  isKioskModeActive: PropTypes.bool,
};

export default Arrow;
