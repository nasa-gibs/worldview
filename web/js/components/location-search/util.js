import CoordinateParser from 'coordinate-parser';
import util from '../../util/util';

/**
 * check for valid coordinates using https://www.npmjs.com/package/coordinate-parser
 * @param {String} position
 * @returns {Bool}
 */
export function isValidCoordinates(position) {
  try {
    const isValid = new CoordinateParser(position);
    return isValid;
  } catch (error) {
    return false;
  }
}

/**
 * Get fixed precision coordinate number
 * @param {String} coordinate
 * @returns {Number} parsed coordinate with fixed precision
 */
export const getCoordinateFixedPrecision = (coordinate) => Number(coordinate.toFixed(4));

/**
 * getFormattedCoordinates
 *
 * @param {String} latitude
 * @param {String} longitude
 *
 * @returns {Array} formattedCoordinates
 */
export const getFormattedCoordinates = (latitude, longitude) => {
  const parsedLatitude = getCoordinateFixedPrecision(latitude);
  const parsedLongitude = getCoordinateFixedPrecision(longitude);

  // format coordinates based on localStorage preference
  const format = util.getCoordinateFormat();
  const coordinates = util.formatCoordinate(
    [parsedLongitude, parsedLatitude],
    format,
  );
  const formattedCoordinates = coordinates.split(',');
  return formattedCoordinates;
};
