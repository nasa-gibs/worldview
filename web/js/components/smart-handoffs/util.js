import googleTagManager from 'googleTagManager';
import util from '../../util/util';

/**
 * Method call to direct the user to Earthdata Search with the necessary URL parameters that
 * encapsulate what the user is intending to try and download data / granules from
 * @param {*} proj
 * @param {*} includeDates
 * @param {*} selectedDate
 * @param {*} selectedCollection
 * @param {*} extentCoords
 * @param {*} showBoundingBox
 */
export default function openEarthDataSearch(proj, includeDates, selectedDate, selectedCollection, extentCoords, showBoundingBox) {
  const { value } = selectedCollection;
  googleTagManager.pushEvent({
    event: 'smart_handoffs_open_eds_data_query',
    selected_collection: value,
  });
  const PROJ_CODES = {
    arctic: '90!0!0!0!0!0',
    geographic: '0.0!-180.0!0!1!0!0,2',
    antarctic: '-90!180!0!2!0!0',
  };
  const { southWest, northEast } = extentCoords;
  const params = {
    q: value,
    p: value,
    m: PROJ_CODES[proj],
    'sb[0]': showBoundingBox ? `${southWest},${northEast}` : undefined,
  };

  if (includeDates) {
    const startDate = `${selectedDate}T00:00:00.000Z`;
    const endDate = `${selectedDate}T23:59:59.999Z`;
    params['[qt]'] = `${startDate},${endDate}`;
  }

  const earthDataSearchURL = `https://search.earthdata.nasa.gov/search/granules${util.toQueryString(params)}`;
  window.open(earthDataSearchURL, '_blank');
}
