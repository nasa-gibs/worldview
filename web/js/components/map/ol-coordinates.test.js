import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import util from '../../util/util';
import OlCoordinates from './ol-coordinates';
import { registerProjections } from '../../fixtures';

const { events } = util;
let container;
let map;

beforeEach(() => {
  container = document.createElement('div');
  document.body.appendChild(container);
  registerProjections();

  map = {
    getEventPixel: jest.fn(),
    getCoordinateFromPixel: () => [0, 0],
  };
  act(() => {
    render(<OlCoordinates />, container);
  });
});

afterEach(() => {
  util.setCoordinateFormat('latlon-dd');
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

test('shows coordinates of (10, 20) when moving the mouse', () => {
  map.getCoordinateFromPixel = () => [10, 20];
  events.trigger('map:mousemove', {}, map, 'EPSG:4326');
  expect(container.innerHTML).toMatchSnapshot();
});

test('shows coordinates of (-160, 20) when moving the mouse over wrapped', () => {
  map.getCoordinateFromPixel = () => [200, 20];
  events.trigger('map:mousemove', {}, map, 'EPSG:4326');
  expect(container.innerHTML).toMatchSnapshot();
});

test('shows 20°00.000\'N, 10°00.000\'E when set to degrees and minutes format', () => {
  util.setCoordinateFormat('latlon-dm');
  map.getCoordinateFromPixel = () => [10, 20];
  events.trigger('map:mousemove', {}, map, 'EPSG:4326');
  expect(container.innerHTML).toMatchSnapshot();
});

test('clears coordinates when mouse moves off the map', () => {
  map.getCoordinateFromPixel = () => [10, 20];
  events.trigger('map:mousemove', {}, map, 'EPSG:4326');
  events.trigger('map:mouseout', {}, map, 'EPSG:4326');
  expect(container.innerHTML).toMatchSnapshot();
});

test('reprojects (0,0) to (-45, 90) for EPSG:3413', () => {
  map.getCoordinateFromPixel = () => [0, 0];
  events.trigger('map:mousemove', {}, map, 'EPSG:3413');
  expect(container.innerHTML).toMatchSnapshot();
});
