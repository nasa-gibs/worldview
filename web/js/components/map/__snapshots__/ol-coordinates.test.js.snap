// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`clears coordinates when mouse moves off the map 1`] = `"<div id=\\"ol-coords-case\\" class=\\"wv-coords-container\\" style=\\"width: 230px;\\"></div>"`;

exports[`reprojects (0,0) to (-45, 90) for EPSG:3413 1`] = `"<div id=\\"ol-coords-case\\" class=\\"wv-coords-container\\" style=\\"width: 230px;\\"><div id=\\"coords-panel\\" class=\\"wv-coords-map wv-coords-map-btn\\"><span class=\\"map-coord\\">90.0000°, -45.0000°</span><div class=\\"map-coord-format\\"><span class=\\"map-coord\\">EPSG:3413</span><div aria-label=\\"Change coordinates format\\" class=\\"coord-btn\\"><i class=\\"coord-switch\\"></i></div></div></div></div>"`;

exports[`shows 20°00.000'N, 10°00.000'E when set to degrees and minutes format 1`] = `"<div id=\\"ol-coords-case\\" class=\\"wv-coords-container\\" style=\\"width: 230px;\\"><div id=\\"coords-panel\\" class=\\"wv-coords-map wv-coords-map-btn\\"><span class=\\"map-coord\\">20.0000°, 10.0000°</span><div class=\\"map-coord-format\\"><span class=\\"map-coord\\">EPSG:4326</span><div aria-label=\\"Change coordinates format\\" class=\\"coord-btn\\"><i class=\\"coord-switch\\"></i></div></div></div></div>"`;

exports[`shows coordinates of (-160, 20) when moving the mouse over wrapped 1`] = `"<div id=\\"ol-coords-case\\" class=\\"wv-coords-container\\" style=\\"width: 230px;\\"><div id=\\"coords-panel\\" class=\\"wv-coords-map wv-coords-map-btn\\"><span class=\\"map-coord\\">20.0000°, -160.0000°</span><div class=\\"map-coord-format\\"><span class=\\"map-coord\\">EPSG:4326</span><div aria-label=\\"Change coordinates format\\" class=\\"coord-btn\\"><i class=\\"coord-switch\\"></i></div></div></div></div>"`;

exports[`shows coordinates of (10, 20) when moving the mouse 1`] = `"<div id=\\"ol-coords-case\\" class=\\"wv-coords-container\\" style=\\"width: 230px;\\"><div id=\\"coords-panel\\" class=\\"wv-coords-map wv-coords-map-btn\\"><span class=\\"map-coord\\">20.0000°, 10.0000°</span><div class=\\"map-coord-format\\"><span class=\\"map-coord\\">EPSG:4326</span><div aria-label=\\"Change coordinates format\\" class=\\"coord-btn\\"><i class=\\"coord-switch\\"></i></div></div></div></div>"`;
